cli/src/commands/mod.rs

```diff
@@ -172,13 +174,80 @@ struct DummyCommandArgs {
     _args: Vec<String>,
 }
 
+fn add_help_cmd_to_subcommands(command: clap::Command, help_cmd: &clap::Command) -> clap::Command {
+    let mut ret = command.clone();
+
+    for subcmd in command.get_subcommands() {
+        ret = ret.mut_subcommand(subcmd.get_name(), |mut subcmd| {
+            if subcmd.has_subcommands() {
+                subcmd = add_help_cmd_to_subcommands(subcmd.clone(), help_cmd);
+                subcmd.disable_help_subcommand(true).subcommand(help_cmd)
+            } else {
+                subcmd
+            }
+        });
+    }
+
+    ret
+}
+
 pub fn default_app() -> clap::Command {
-    Command::augment_subcommands(Args::command())
+    let app = Command::augment_subcommands(Args::command()).disable_help_subcommand(true);
+    add_help_cmd_to_subcommands(
+        app.clone(),
+        app.find_subcommand("help")
+            .expect("JJ has the help command"),
+    )
+}
+
+fn has_help_subcmd(matches: &clap::ArgMatches) -> bool {
+    let mut curr_matches = matches;
+
+    while let Some((subcmd_name, matches)) = curr_matches.subcommand() {
+        if subcmd_name == "help" {
+            return true;
+        }
+
+        curr_matches = matches;
+    }
+
+    false
+}
+
+fn get_help_arg_cmd(matches: &clap::ArgMatches) -> Vec<String> {
+    let mut curr_matches = matches;
+
+    let mut ret = vec![];
+
+    while let Some((subcmd_name, matches)) = curr_matches.subcommand() {
+        if subcmd_name != "help" {
+            ret.push(subcmd_name.to_string());
+        } else {
+            let cmd = Command::from_arg_matches(curr_matches).unwrap();
+            if let Command::Help(args) = cmd {
+                ret.extend(args.command);
+            } else {
+                unreachable!()
+            }
+            break;
+        }
+
+        curr_matches = matches;
+    }
+
+    ret
 }
 
 #[instrument(skip_all)]
 pub fn run_command(ui: &mut Ui, command_helper: &CommandHelper) -> Result<(), CommandError> {
-    let subcommand = Command::from_arg_matches(command_helper.matches()).unwrap();
+    let subcommand = if has_help_subcmd(command_helper.matches()) {
+        Command::Help(help::HelpArgs {
+            command: get_help_arg_cmd(command_helper.matches()),
+        })
+    } else {
+        Command::from_arg_matches(command_helper.matches()).unwrap()
+    };
```

Comment:
Do you think it would be simpler to remove `has_help_subcmd()` and let `get_help_arg_cmd()` return `Option<Vec<String>>`?